Implement more smart pointers in WebProcess.h/cpp (part 2)

lopezlozoya opened the pull request:
#### 1062357632e222dbb55a4b21a98f91bebbf49a02
<pre>
Implement more smart pointers in WebProcess.h/cpp (part 2)
<a href="https://bugs.webkit.org/show_bug.cgi?id=287895">https://bugs.webkit.org/show_bug.cgi?id=287895</a>

Reviewed by Ryosuke Niwa.

Deploy more smart pointers in WebProcess (part 2).

* Source/WebCore/workers/service/context/SWContextManager.h:
(WebCore::SWContextManager::protectedConnection const):
* Source/WebCore/workers/shared/context/SharedWorkerContextManager.h:
(WebCore::SharedWorkerContextManager::protectedConnection const):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::protectedServiceWorkerConnection):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
(WebKit::WebProcess::registerURLSchemeAsCORSEnabled):
(WebKit::WebProcess::createWebPage):
(WebKit::WebProcess::shouldTerminate):
(WebKit::WebProcess::dispatchMessage):
(WebKit::WebProcess::removeWebFrame):
(WebKit::WebProcess::userGestureTokenIdentifier):
(WebKit::WebProcess::userGestureTokenDestroyed):
(WebKit::WebProcess::backgroundResponsivenessPing):
(WebKit::WebProcess::handleInjectedBundleMessage):
(WebKit::WebProcess::ensureNetworkProcessConnection):
(WebKit::WebProcess::protectedNetworkProcessConnection):
(WebKit::WebProcess::logDiagnosticMessageForNetworkProcessCrash):
(WebKit::WebProcess::networkProcessConnectionClosed):
(WebKit::WebProcess::gpuProcessConnectionDidBecomeUnresponsive):
(WebKit::WebProcess::reloadExecutionContextsForOrigin):
(WebKit::WebProcess::prepareToSuspend):
(WebKit::WebProcess::processDidResume):
(WebKit::WebProcess::sendPrewarmInformation):
(WebKit::WebProcess::prefetchDNS):
(WebKit::WebProcess::establishRemoteWorkerContextConnectionToNetworkProcess):
(WebKit::WebProcess::revokeUserMediaDeviceSandboxExtensions):
(WebKit::WebProcess::setAppBadge):
(WebKit::WebProcess::setClientBadge):
(WebKit::WebProcess::setUseGPUProcessForMedia):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: <a href="https://commits.webkit.org/290775@main">https://commits.webkit.org/290775@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b8ad0e6f13dd7656b9ac963f462123217e85b6f3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91046 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10589 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/57 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96066 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41832 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93098 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10984 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18907 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69980 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27505 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94047 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8384 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82516 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50306 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8150 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/43 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40956 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78472 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/92 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98083 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18244 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18503 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78323 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78219 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19329 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22687 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/54 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18250 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/23607 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17994 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21460 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19778 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -826,7 +826,7 @@ void WebProcess::registerURLSchemeAsDisplayIsolated(const String& urlScheme) con
 {
     LegacySchemeRegistry::registerURLSchemeAsCORSEnabled(urlScheme);
-    ensureNetworkProcessConnection().connection().send(Messages::NetworkConnectionToWebProcess::RegisterURLSchemesAsCORSEnabled({ urlScheme }), 0);
+    ensureProtectedNetworkProcessConnection()->protectedConnection()->send(Messages::NetworkConnectionToWebProcess::RegisterURLSchemesAsCORSEnabled({ urlScheme }), 0);
```
What is the error from the analyzer here? I doubt we need `ensureProtectedNetworkProcessConnection`. I bet `ensureNetworkProcessConnection().protectedConnection()->` should be enough to fix it because `protectedConnection` is a trivial function.

lopezlozoya replied:
Will fix this when I fix the merge conflict.

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1283,9 +1283,9 @@ Ref<NetworkProcessConnection> WebProcess::ensureProtectedNetworkProcessConnectio
 void WebProcess::logDiagnosticMessageForNetworkProcessCrash()
 {
-    WebCore::Page* page = nullptr;
+    RefPtr<WebCore::Page> page = nullptr;
```
Just `RefPtr<WebCore::Page> page;` should be enough. We don't need the `= nullptr` because RefPtr is default-initialized to `nullptr`.

lopezlozoya replied:
Thanks for the advice, I'll change it.

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1337,26 +1337,26 @@ void WebProcess::networkProcessConnectionClosed(NetworkProcessConnection* connec
     m_webLoaderStrategy->networkProcessCrashed();
     m_webSocketChannelManager.networkProcessCrashed();
-    m_broadcastChannelRegistry->networkProcessCrashed();
+    Ref { m_broadcastChannelRegistry }->networkProcessCrashed();
```
Can we mark `m_broadcastChannelRegistry` as `const` instead? It seems like it's never re-assigned.

lopezlozoya replied:
I tried changing it and had some compile time issues, I'll try to address it once I rebase this PR.

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1337,26 +1337,26 @@ void WebProcess::networkProcessConnectionClosed(NetworkProcessConnection* connec
     }
 
-    m_cacheStorageProvider->networkProcessConnectionClosed();
+    Ref { m_cacheStorageProvider }->networkProcessConnectionClosed();
```
Can we mark `m_cacheStorageProvider` as `const` instead? It seems like it's never re-assigned.

lopezlozoya replied:
I'll fix this.

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1337,26 +1337,26 @@ void WebProcess::networkProcessConnectionClosed(NetworkProcessConnection* connec
     if (m_fileSystemStorageConnection) {
-        m_fileSystemStorageConnection->connectionClosed();
+        RefPtr { m_fileSystemStorageConnection }->connectionClosed();
         m_fileSystemStorageConnection = nullptr;
```
`if (RefPtr fileSystemStorageConnection = m_fileSystemStorageConnection) {`
`       fileSystemStorageConnection->connectionClosed();`
`       m_fileSystemStorageConnection = nullptr;`
`}`

lopezlozoya replied:
I'll address this, thanks.

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1583,7 +1583,7 @@ void WebProcess::deleteWebsiteDataForOrigin(OptionSet<WebsiteDataType> websiteDa
 {
     for (auto& page : m_pageMap.values()) {
-        if (auto* corePage = page->corePage())
+        if (RefPtr corePage = page->protectedCorePage())
```
`if (RefPtr corePage = page->corePage())`

lopezlozoya replied:
Thanks

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2044,7 +2044,7 @@ void WebProcess::prefetchDNS(const String& hostname)
 
     if (m_dnsPrefetchedHosts.add(hostname).isNewEntry)
-        ensureNetworkProcessConnection().connection().send(Messages::NetworkConnectionToWebProcess::PrefetchDNS(hostname), 0);
+        ensureProtectedNetworkProcessConnection()->protectedConnection()->send(Messages::NetworkConnectionToWebProcess::PrefetchDNS(hostname), 0);
```
Same comment as above. I suspect we don't need `ensureProtectedNetworkProcessConnection()`.

lopezlozoya replied:
Got it, I was unsure if need to used protected for ensureNetworkProcessConnection

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2092,11 +2092,11 @@ void WebProcess::establishRemoteWorkerContextConnectionToNetworkProcess(RemoteWo
     case RemoteWorkerType::SharedWorker:
         SharedWorkerContextManager::singleton().setConnection(WebSharedWorkerContextManagerConnection::create(WTFMove(ipcConnection), WTFMove(site), pageGroupID, webPageProxyID, pageID, store, WTFMove(initializationData)));
-        SharedWorkerContextManager::singleton().connection()->establishConnection(WTFMove(completionHandler));
+        RefPtr { SharedWorkerContextManager::singleton().connection() }->establishConnection(WTFMove(completionHandler));
```
Since this is used in multiple places, let's add a `protectedConnection()` function and use that instead.

lopezlozoya replied:
I'll add that method, thank you.

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2188,7 +2188,7 @@ void WebProcess::revokeUserMediaDeviceSandboxExtensions(const Vector<String>& ex
     
     if (m_machBootstrapExtension)
-        m_machBootstrapExtension->revoke();
+        RefPtr { m_machBootstrapExtension }->revoke();
```
`if (RefPtr machBootstrapExtension = m_machBootstrapExtension)`
`       machBootstrapExtension->revoke();`

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2224,17 +2224,17 @@ void WebProcess::setAppBadge(std::optional<WebPageProxyIdentifier> pageIdentifie
     if (DeprecatedGlobalSettings::builtInNotificationsEnabled()) {
         if (m_sessionID)
-            ensureNetworkProcessConnection().connection().send(Messages::NotificationManagerMessageHandler::SetAppBadge({ origin, badge }), m_sessionID->toUInt64());
+            ensureProtectedNetworkProcessConnection()->protectedConnection()->send(Messages::NotificationManagerMessageHandler::SetAppBadge({ origin, badge }), m_sessionID->toUInt64());
```
Same comment as above. I suspect we don't need ensureProtectedNetworkProcessConnection().

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2092,11 +2092,11 @@ void WebProcess::establishRemoteWorkerContextConnectionToNetworkProcess(RemoteWo
     case RemoteWorkerType::ServiceWorker:
         SWContextManager::singleton().setConnection(WebSWContextManagerConnection::create(WTFMove(ipcConnection), WTFMove(site), serviceWorkerPageIdentifier, pageGroupID, webPageProxyID, pageID, store, WTFMove(initializationData)));
-        SWContextManager::singleton().connection()->establishConnection(WTFMove(completionHandler));
+        RefPtr { SWContextManager::singleton().connection() }->establishConnection(WTFMove(completionHandler));
```
Let's use `SharedWorkerContextManager::singleton().protectedConnection()->` here as well.

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1318,13 +1318,13 @@ void WebProcess::networkProcessConnectionClosed(NetworkProcessConnection* connec
             continue;
         
-        if (auto* existingIDBConnectionToServer = connection->existingIDBConnectionToServer()) {
+        if (RefPtr existingIDBConnectionToServer = RefPtr { connection->existingIDBConnectionToServer() }) {
```
The construction of the RefPtr is implicit so we can just do: `if (RefPtr existingIDBConnectionToServer = connection->existingIDBConnectionToServer())`.

lopezlozoya replied:
Thank you, I'll address this and update the PR.

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -487,22 +487,22 @@ void WebProcess::initializeWebProcess(WebProcessCreationParameters&& parameters,
 #endif
-        memoryPressureHandler.setMemoryKillCallback([this] () {
+        memoryPressureHandler.setMemoryKillCallback([this, protectedThis = Ref { *this }] () {
             WebCore::logMemoryStatistics(LogMemoryStatisticsReason::OutOfMemoryDeath);
```
Let's put `RefPtr parentProcessConnection = this->parentProcessConnection();` below this line and then use this local variable in the below two instances instead of repeatedly calling `protectedParentProcessConnection()`.

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -487,22 +487,22 @@ void WebProcess::initializeWebProcess(WebProcessCreationParameters&& parameters,
 #endif
-        memoryPressureHandler.setMemoryPressureStatusChangedCallback([this]() {
+        memoryPressureHandler.setMemoryPressureStatusChangedCallback([this, protectedThis = Ref { *this }]() {
             if (parentProcessConnection())
```
`if (RefPtr parentProcessConnection = this->parentProcessConnection())`
`    parentProcessConnection->send(Messages::WebProcessProxy::MemoryPressureStatusChanged(MemoryPressureHandler::singleton().memoryPressureStatus()), 0);`

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1318,13 +1318,13 @@ void WebProcess::networkProcessConnectionClosed(NetworkProcessConnection* connec
     }
 
     for (auto& page : m_pageMap.values()) {
-        auto idbConnection = page->corePage()->optionalIDBConnection();
```
Add `RefPtr corePage = page->corePage();` and then use this local variable below instead of calling `protectedCorePage()` twice.

RupinMittal commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2092,11 +2092,11 @@ void WebProcess::establishRemoteWorkerContextConnectionToNetworkProcess(RemoteWo
     case RemoteWorkerType::ServiceWorker:
         SWContextManager::singleton().setConnection(WebSWContextManagerConnection::create(WTFMove(ipcConnection), WTFMove(site), serviceWorkerPageIdentifier, pageGroupID, webPageProxyID, pageID, store, WTFMove(initializationData)));
-        SWContextManager::singleton().connection()->establishConnection(WTFMove(completionHandler));
+        SharedWorkerContextManager::singleton().protectedConnection()->establishConnection(WTFMove(completionHandler));
```
Ah my last comment was about using `protectedConnection()`. But we should still use `SWContextManager` here instead of switching to `SharedWorkerContextManager`.

lopezlozoya replied:
Good catch, that was a typo, thank you.

RupinMittal reviewed (commented):
Almost there.

lopezlozoya committed:
Implement more smart pointers in WebProcess.h/cpp (part 2)
https://bugs.webkit.org/show_bug.cgi?id=287895

Reviewed by Ryosuke Niwa.

Deploy more smart pointers in WebProcess (part 2).

* Source/WebCore/workers/service/context/SWContextManager.h:
(WebCore::SWContextManager::protectedConnection const):
* Source/WebCore/workers/shared/context/SharedWorkerContextManager.h:
(WebCore::SharedWorkerContextManager::protectedConnection const):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::protectedServiceWorkerConnection):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
(WebKit::WebProcess::registerURLSchemeAsCORSEnabled):
(WebKit::WebProcess::createWebPage):
(WebKit::WebProcess::shouldTerminate):
(WebKit::WebProcess::dispatchMessage):
(WebKit::WebProcess::removeWebFrame):
(WebKit::WebProcess::userGestureTokenIdentifier):
(WebKit::WebProcess::userGestureTokenDestroyed):
(WebKit::WebProcess::backgroundResponsivenessPing):
(WebKit::WebProcess::handleInjectedBundleMessage):
(WebKit::WebProcess::ensureNetworkProcessConnection):
(WebKit::WebProcess::protectedNetworkProcessConnection):
(WebKit::WebProcess::logDiagnosticMessageForNetworkProcessCrash):
(WebKit::WebProcess::networkProcessConnectionClosed):
(WebKit::WebProcess::gpuProcessConnectionDidBecomeUnresponsive):
(WebKit::WebProcess::reloadExecutionContextsForOrigin):
(WebKit::WebProcess::prepareToSuspend):
(WebKit::WebProcess::processDidResume):
(WebKit::WebProcess::sendPrewarmInformation):
(WebKit::WebProcess::prefetchDNS):
(WebKit::WebProcess::establishRemoteWorkerContextConnectionToNetworkProcess):
(WebKit::WebProcess::revokeUserMediaDeviceSandboxExtensions):
(WebKit::WebProcess::setAppBadge):
(WebKit::WebProcess::setClientBadge):
(WebKit::WebProcess::setUseGPUProcessForMedia):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: https://commits.webkit.org/290775@main
